compiler/lib/inline.ml

```diff
@@ -21,129 +21,457 @@
+      (fun pc () ->
+        let block = Addr.Map.find pc p.blocks in
+        Freevars.iter_block_free_vars add_dep block;
+        List.iter
+          ~f:(fun i ->
+            match i with
+            | Let (f, Closure _) -> add_dep f
+            | _ -> ())
+          block.body)
+      pc
+      p.blocks
+      ()
+  in
+  Var.Hashtbl.iter (fun f (_, (pc, _), _) -> traverse p f pc) closures;
+  Var.Hashtbl.fold (fun f s m -> Var.Map.add f !s m) deps Var.Map.empty
+
+module Var_SCC = Strongly_connected_components.Make (Var)
+
+let visit_closures p ~live_vars f acc =
+  let closures = collect_closures p in
+  let deps = collect_deps p closures in
+  let scc = Var_SCC.connected_components_sorted_from_roots_to_leaf deps in
+  let f' recursive acc g =
+    let params, cont, enclosing_function = Var.Hashtbl.find closures g in
+    f ~recursive ~enclosing_function ~current_function:(Some g) ~params ~cont acc
+  in
+  let acc =
+    Array.fold_left
+      scc
+      ~f:(fun acc group ->
+        match group with
+        | Var_SCC.No_loop g -> f' false acc g
+        | Has_loop l ->
+            let set = Var.Set.of_list l in
+            let deps' =
+              List.fold_left
+                ~f:(fun deps' g ->
+                  Var.Map.add
+                    g
+                    (if live_vars.(Var.idx g) > 1
+                     then Var.Set.empty
+                     else Var.Set.inter (Var.Map.find g deps) set)
+                    deps')
+                ~init:Var.Map.empty
+                l
+            in
+            let scc = Var_SCC.connected_components_sorted_from_roots_to_leaf deps' in
+            Array.fold_left
+              scc
+              ~f:(fun acc group ->
+                match group with
+                | Var_SCC.No_loop g -> f' true acc g
+                | Has_loop l -> List.fold_left ~f:(fun acc g -> f' true acc g) ~init:acc l)
+              ~init:acc)
+      ~init:acc
+  in
+  f
+    ~recursive:false
+    ~enclosing_function:None
+    ~current_function:None
+    ~params:[]
+    ~cont:(p.start, [])
+    acc
+
+(****)
+
+module SCC = Strongly_connected_components.Make (Addr)
+
+let blocks_in_loop p pc =
+  let g =
+    Code.traverse
+      { fold = Code.fold_children }
+      (fun pc g ->
+        Addr.Map.add pc (Code.fold_children p.blocks pc Addr.Set.add Addr.Set.empty) g)
+      pc
+      p.blocks
+      Addr.Map.empty
+  in
+  let scc = SCC.component_graph g in
+  Array.fold_left
+    ~f:(fun s (c, _) ->
+      match c with
+      | SCC.No_loop _ -> s
+      | Has_loop l -> List.fold_left ~f:(fun s x -> Addr.Set.add x s) l ~init:s)
+    ~init:Addr.Set.empty
+    scc
+
+(****)
+
+type 'a cache = 'a option ref
+
+type info =
+  { f : Var.t
+  ; params : Var.t list
+  ; cont : Code.cont
+  ; enclosing_function : Var.t option
+  ; recursive : bool
+  ; loops : bool cache
+  ; body_size : int cache
+  ; full_size : int cache
+  ; closure_count : int cache
+  ; init_code : int cache
```

Comment:
Could you add a comment to some of these fields explain what there are for ? 
At least for `init_code`, `interesting_params`, `full_size`

Reply:
I have added a comment to the functions that fill these fields.